File: src/main/java/bad/robot/radiate/monitor/RandomStatus.java
package bad.robot.radiate.monitor;

import bad.robot.radiate.Activity;
import bad.robot.radiate.Progress;
import bad.robot.radiate.Status;

import java.security.SecureRandom;
import java.util.Random;

import static bad.robot.radiate.Activity.*;
import static bad.robot.radiate.Activity.Error;
import static bad.robot.radiate.Status.*;

public class RandomStatus extends ThreadSafeObservable implements MonitoringTask {

    private static final Random random = new SecureRandom();
    private static final Status[] statuses = new Status[]{Ok, Ok, Ok, Ok, Ok, Ok, Ok, Ok, Ok, Broken, Unknown};
    private static final Activity[] activities = new Activity[]{Busy, Error, Idle, Progressing};

    @Override
    public void run() {
        Status status = randomStatus();
        notifyObservers(randomState(), randomProgress());
        notifyObservers(status);
        if (status == Broken)
            notifyObservers(new RuntimeException("Example problem"));
    }

    private Progress randomProgress() {
        return new Progress(random.nextInt(100) + 1, 100);
    }

    private static Status randomStatus() {
        return statuses[random.nextInt(statuses.length)];
    }

    private static Activity randomState() {
        return activities[random.nextInt(activities.length)];
    }

    public String toString() {
        return "randomly passing build monitoring";
    }
}

File: src/main/scala/bad/robot/radiate/monitor/RandomStatus.scala
package bad.robot.radiate.monitor

import java.security.SecureRandom

import bad.robot.radiate.monitor.RandomStatus._
import bad.robot.radiate._
import bad.robot.radiate._
import activity._

object RandomStatus {

  private val random = new SecureRandom
  private val statuses = Array(Ok, Ok, Ok, Ok, Ok, Ok, Ok, Ok, Ok, Broken, Unknown)
  private val activities = Array(Busy, Error, Idle, Progressing)

  def randomProgress: Progress = {
    new Progress(random.nextInt(100) + 1, 100)
  }

  def randomStatus: Status = {
    statuses(random.nextInt(statuses.length))
  }

  private def randomActivity: Activity = {
    activities(random.nextInt(activities.length))
  }
}

class RandomStatus extends ThreadSafeObservable with MonitoringTask {
  def run {
    val activity = randomActivity
    val status = randomStatus
    notifyObservers(activity, if (activity == Progressing) randomProgress else new NullProgress)
    notifyObservers(status)
    if (status == Broken)
      notifyObservers(new RuntimeException("Example problem"))
  }

  override def toString: String = "randomly passing build monitoring"
}